Avoid formatting the body of a define

mihaipopescu opened the pull request:
Changed the way we are tokenizing PP_IGNORE. Previously we were forcing every `define` token to be PP_IGNORE but that didn't worked for multiline defines.

With this we're instead tokenizing the body of a define with PP_IGNORE avoiding doing any extra formatting for these tokens.

mihai-unity committed:
Test case for PP_IGNORE fix

mihai-unity committed:
formatting: tabs -> spaces

mihai-unity committed:
added `pp_ignore_define_body` option to properly tokenize PP_IGNORE without having to force with set

mihai-unity committed:
activate the new option for staging tests and make the test as a regression
because we will fix it in this branch.

mihai-unity committed:
When we are detecting CT_MACRO all the following tokens are actually forming the macro body.
In this state, we're just going to tokenize all chunks (except NL_CONT) to ignore.
When we have a CT_MACRO_FUNC we have to wait until we see a PAREN_CLOSE
then start entering the CT_PP_IGNORE state.

mihai-unity committed:
After the tokenization is done correctly for CT_PP_IGNORE we should avoid doing
any formatting on these tokens.

mihai-unity committed:
Comments should not be ignored if they appear part of the define body
Fixes 10048-UNI-1335.cpp regression test.

mihai-unity committed:
Accepting the space before nl_cont becaue we have `sp_before_nl_cont=add`

scottbilas commented on src/indent.cpp:
```diff
@@ -2105,7 +2105,7 @@ void indent_text(void)
             }
 
             /* [Unity-only hack] Do nothing. We're using 'define' as CT_PP_IGNORE to avoid reformatting of complex and weird macros. Likewise leave indentation alone because probably also weird. */
```
comment is out of date. and also should we just look at using pp_ignore_define_body around this and no longer consider it a "unity hack"?

scottbilas commented on src/space.cpp:
```diff
@@ -129,6 +129,11 @@ static argval_t do_space(chunk_t *first, chunk_t *second, int& min_sp, bool comp
    }
+   if ((first->type == CT_PP_IGNORE) && (second->type == CT_PP_IGNORE))
+   {
+	   log_rule("PP_IGNORE");
```
needs some kind of comment about how CT_PP_IGNORE doesn't make sense as something we need to adjust spacing for. the point of it is to ignore blocks of preprocessor code, and definitely don't want the default fallthrough behavior.

scottbilas commented on src/tokenize.cpp:
```diff
@@ -1228,13 +1234,13 @@ bool parse_word(tok_ctx& ctx, chunk_t& pc, bool skipcheck)
-		 if (pc.type == CT_PP_IGNORE && !cpd.in_preproc)
-			 pc.type = find_keyword_type(pc.text(), pc.str.size(), false);
+         /* Special pattern: if we're trying to redirect a preprocessor directive to PP_IGNORE,
+             then ensure we're actually part of a preprocessor before doing the swap, or we'll
```
extra space?

mihai-unity committed:
Removed unity-only hack for avoiding indentation by walking backwards on PREPROC_BODY until PP_IGNORE is found
because now the preprocessor body is properly tokenized as PP_IGNORE instead of the #define token.

mihai-unity committed:
indentation: minor changes and improved comments

mihai-unity committed:
The code from output.cpp doesn't allow modifying the colums of tokens
which are following after a PP_IGNORE token to avoid bad formatting.
This test is one corner case where the result is a misaligninment which is accepted
for now and left for the user to fix it manually.
